File: portal/src/utils/words.ts
/**
 * ONE definition of "a word" for the client, mirroring `backend/src/core/words.py`. Two
 * surfaces count words — project title (max 8), delete reason (5–50) — and client/server
 * must split identically, or a message that passes in the browser gets refused by the API
 * with no way to tell what it disliked (counter reads `8/8`, Create enabled, API 422s).
 * Python: `len(value.split())`. TypeScript: `value.split(PY_WHITESPACE).filter(Boolean).length`
 * — `\s` is CLOSE to Python's whitespace class but not equal (see `PY_WHITESPACE` below), so
 * the class is written out; `.filter(Boolean)` drops what Python discards for free. Verified
 * equivalent on empty/whitespace/tab/newline/CRLF/leading-trailing/NBSP/ideographic-space —
 * `words.test.ts` pins the list, `test_project_name_words.py` pins its mirror, because "same
 * rule both sides" is only a fact while something checks it.
 */

/**
 * Python's whitespace set, written out, because `\s` is NOT it — they differ on six code
 * points (Python's `isspace()` adds `U+001C`–`U+001F`/`U+0085` NEL; `\s` adds `U+FEFF` BOM).
 * Nobody types these on purpose, but a pasted spreadsheet title or a BOM'd UTF-8 file
 * carries them, and the point of this module is that the count a user watches cannot
 * disagree with the validator that refuses them. `.trim()` is deliberately unused too: it
 * trims by the JS set, so a leading BOM would vanish here but count as a word on the server.
 */
// The control characters below are the POINT, not a typo. eslint's no-control-regex exists
// because one in a regex is normally accidental; U+001C-U+001F are in Python's whitespace
// set, and omitting them is exactly what made the two counters disagree. The exhaustive
// sweep in `words.test.ts` is what keeps the two sides honest.
const PY_WHITESPACE =
  // eslint-disable-next-line no-control-regex
  /[\t\n\v\f\r \u001c-\u001f\u0085\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]+/

/** How many words `value` contains, by the shared rule above. */
export function countWords(value: string): number {
  return value.split(PY_WHITESPACE).filter(Boolean).length
}

/** The project title cap — mirrors `MAX_PROJECT_NAME_WORDS` in
 *  `backend/src/db/models/project.py`. */
export const MAX_PROJECT_NAME_WORDS = 8

/** The project description bounds — mirrors `MIN_PROJECT_DESCRIPTION_WORDS` /
 *  `MAX_PROJECT_DESCRIPTION_WORDS` in `backend/src/db/models/project.py`. Unlike the title,
 *  description has a MINIMUM: a one-line description embeds into a single vector for
 *  semantic search, so one too short to say anything embeds to nothing worth matching. */
export const MIN_PROJECT_DESCRIPTION_WORDS = 15
export const MAX_PROJECT_DESCRIPTION_WORDS = 120

/** The delete-reason bounds (#158 §13.2). */
export const MIN_DELETE_REASON_WORDS = 5
export const MAX_DELETE_REASON_WORDS = 50

/** THE COLUMN'S PASTE BACKSTOP, not the rule anybody is told about — that is the word count.
 *  Mirrors the server's own `clean_deletion_reason` cap, and lives here because BOTH deletes
 *  now need it: the citizen's project delete and the administrator's app delete. It was a
 *  private constant in `ProjectDeleteDialog` while only one of them existed. */
export const MAX_DELETE_REASON_CHARS = 2000

File: backend/src/core/words.py
"""ONE definition of "a word", shared by every server-side word limit.

Two surfaces count words (#158): the project title (§14, max 8) and the reason someone
gives for deleting a project (§13.2, 5–50). The issue is explicit about the hazard:

>   Define "word" once and share it. Client and server must split identically, or a
>   message that passes in the browser gets refused by the API.

So this is the server's single definition, and `portal/src/utils/words.ts` is the client's.
They are deliberately small, and deliberately equivalent:

    Python   len(value.split())
    TypeScript   value.split(PY_WHITESPACE).filter(Boolean).length

`str.split()` with no argument is the exact behaviour we want, and is not the same as
`split(" ")`: it splits on RUNS of arbitrary Unicode whitespace and discards empty tokens,
so a double space, a tab, a newline, a CRLF, a non-breaking space (U+00A0) and an
ideographic space (U+3000) all behave the way a reader would expect.

THE CLIENT SPELLS THE WHITESPACE SET OUT, and that is not fussiness: JavaScript's `\\s` is
NOT this set. Sweeping all 1,114,112 code points found six disagreements — Python's
`str.isspace()` includes the separators U+001C-U+001F and U+0085 (NEL), which `\\s` does not,
and `\\s` matches U+FEFF (the BOM), which Python does not. Nobody types those on purpose, but
a title pasted out of a spreadsheet, a mainframe export or a BOM'd file carries them, and
the whole point of this pair is that the counter someone watches cannot disagree with the
validator that refuses them. `words.ts` therefore writes the class out as `PY_WHITESPACE`
and skips `.trim()`, which trims by the JS set; `.filter(Boolean)` drops the empty
leading/trailing tokens Python discards for free.

Verified equivalent on: `""`, `"   "`, `"one"`, `"a  b"`, `"a\\tb"`, `"a\\nb"`, `"a\\r\\nb"`,
`" lead and trail "`, `"a\\u00a0b"`, `"a\\u3000b"`. Both sides have a test pinning that list,
because the whole point is that they cannot drift apart.

Counting words rather than characters is the issue's choice, not ours to relitigate here:
a title is a name, and "about 6 to 8 words" is what a person can act on. The one thing a
word rule owes them in return is that the browser and the API agree, which is this module.
"""

from __future__ import annotations


def count_words(value: str) -> int:
    """How many words `value` contains, by the shared rule above.

    `str.split()` with NO argument, never `split(" ")` — the latter would count a double
    space as an extra empty word and disagree with the client.
    """
    return len(value.split())
